- First QC: ```sh batch_qc.sh -i /data/public/jkoubele/fli_dr_mice/FASTQ -o /data/public/jkoubele/fli_dr_mice/QC_before_trimming```
- Detecting adapters: ```sh batch_detect_adapers.sh -i /data/public/jkoubele/fli_dr_mice/FASTQ -o /data/public/jkoubele/fli_dr_mice/detected_adapters```
- Aggregate adapters: ```sh run_aggregate_adapters.sh -i /data/public/jkoubele/fli_dr_mice/detected_adapters -o /data/public/jkoubele/fli_dr_mice/aggregated_adapters```
   After that, check if the adapters were consistent across samples (the output will contain flag whether the adapters were detected successfully.)
- Trimming reads: ```sh batch_trimming.sh -i /data/public/jkoubele/fli_dr_mice/FASTQ -o /data/public/jkoubele/fli_dr_mice/FASTQ_trimmed -a /data/public/jkoubele/fli_dr_mice/aggregated_adapters```
- QC after trimming: ```sh batch_qc.sh -i /data/public/jkoubele/fli_dr_mice/FASTQ_trimmed/ -o /data/public/jkoubele/fli_dr_mice/QC_after_trimming```
- Alignment: ```sh batch_align.sh -i /data/public/jkoubele/fli_dr_mice/FASTQ_trimmed/ -o /data/public/jkoubele/fli_dr_mice/BAM_before_deduplication -g /data/public/jkoubele/reference_genomes/GRCm39```
- Infer strandedness: ```sh run_infer_strandedness.sh -i /data/public/jkoubele/fli_dr_mice/BAM_before_deduplication -o /data/public/jkoubele/fli_dr_mice/strandedness_info```
- UMI deduplication: ```sh batch_deduplicate_umi.sh -i /data/public/jkoubele/fli_dr_mice/BAM_before_deduplication/ -o /data/public/jkoubele/fli_dr_mice/BAM```
- Feature counts: ```sh batch_feature_counts.sh -i /data/public/jkoubele/fli_dr_mice/BAM -o /data/public/jkoubele/fli_dr_mice/feature_counts -g /data/public/jkoubele/reference_genomes/GRCm39 -a Mus_musculus.GRCm39.112.gtf -s 2```
- Compute coverage: ```sh batch_compute_coverage.sh -i /data/public/jkoubele/fli_dr_mice/BAM -o /data/public/jkoubele/fli_dr_mice/coverage -g /data/public/jkoubele/reference_genomes/GRCm39 -f Mus_musculus.GRCm39.dna.primary_assembly.fa.fai -s 2```
